core/include/detray/core/detail/context_aware_store.hpp

```diff
@@ -0,0 +1,360 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022-2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/core/detail/container_buffers.hpp"
+#include "detray/core/detail/container_views.hpp"
+#include "detray/core/detail/data_context.hpp"
+#include "detray/definitions/detail/indexing.hpp"
+#include "detray/definitions/detail/qualifiers.hpp"
+
+// Vecmem include(s)
+#include <vecmem/memory/memory_resource.hpp>
+
+// System include(s)
+#include <type_traits>
+
+namespace detray {
+
+/// @brief Wraps a jagged vector-like container and implements a data store
+/// around it.
+///
+/// @tparam T The type of the collection data, e.g. transforms
+/// @tparam container_t The type of container to use for the data collection.
+/// @tparam context_t the context with which to retrieve the correct data.
+template <typename T, template <typename...> class container_t,
+          typename context_t>
+class context_aware_store {
+    public:
+    using base_type = container_t<T>;
+    using vector_type = typename base_type::value_type;
+    using context_type = context_t;
+
+    // TODO: do we need this?
+    /*
+    /// How to find data in the store
+    /// @{
+    using link_type = dindex;
+    using single_link = dindex;
+    using range_link = dindex_range;
+    /// @}
+    */
+
+    /// Vecmem view types
+    using view_type = detail::get_view_t<container_t<T>>;
+    /*
+     TODO: uncommenting the next line results in compilation problems:
```

Comment:
It could be that some of the type traits used there for jagged vectors clashes with those for ```vector<T>```. So we probably need to restrict ```T```